File: src/main/java/com/example/DemoWebSocketHandler.java
package com.example;

import org.springframework.web.reactive.socket.WebSocketHandler;
import org.springframework.web.reactive.socket.WebSocketMessage;
import org.springframework.web.reactive.socket.WebSocketSession;
import reactor.core.publisher.Flux;
import reactor.core.publisher.Mono;

/**
 * @author Rob Winch
 */
public class DemoWebSocketHandler implements WebSocketHandler {
	@Override
	public Mono<Void> handle(WebSocketSession session) {
		Flux<WebSocketMessage> message = session.getHandshakeInfo()
			.getPrincipal()
			.map(it -> {
				return session.textMessage(it.getName());
			})
			.switchIfEmpty(Mono.fromSupplier(() -> session.textMessage("ERROR")))
			.concatWith(session.receive().then(Mono.empty()));

		return session.send(message);
	}
}

File: src/main/kotlin/com/example/DemoWebSocketHandler.kt
package com.example

import org.springframework.web.reactive.socket.WebSocketHandler
import org.springframework.web.reactive.socket.WebSocketSession
import reactor.core.publisher.Mono

/**
 * @author Sola
 */
object DemoWebSocketHandler : WebSocketHandler {

    override fun handle(session: WebSocketSession): Mono<Void> {
        return session.handshakeInfo.principal.map {
            session.textMessage(it.toString())
        }.switchIfEmpty(Mono.just(session.textMessage("You shouldn't be seeing this.")))
            .concatWith(session.receive().then(Mono.empty()))
            .let(session::send)
    }

}
